sklearn/impute/_knn.py

```diff
@@ -0,0 +1,290 @@
+
+        # Final imputation
+        imputed = np.ma.average(donors, axis=1, weights=weight_matrix)
+        X_col[receivers_idx] = imputed.data
+
+    def fit(self, X, y=None):
+        """Fit the imputer on X.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features, )
+            Input data, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        self : object
+        """
+
+        # Check data integrity and calling arguments
+        if not is_scalar_nan(self.missing_values):
+            force_all_finite = True
+        else:
+            force_all_finite = "allow-nan"
+            if self.metric not in _NAN_METRICS and not callable(self.metric):
+                raise ValueError(
+                    "The selected metric does not support NaN values")
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+
+        # Check if % missing in any samples > sample_max_missing
+        mask = _get_mask(X, self.missing_values)
+        if np.any(mask.sum(axis=0) > (X.shape[0] * self.sample_max_missing)):
+            raise ValueError("Some columns have more than {}% missing values"
+                             .format(self.sample_max_missing * 100))
+
+        self.weights = _check_weights(self.weights)
+        self.fit_X_ = X[
+            mask.sum(axis=1) <= (mask.shape[1] * self.feature_max_missing)]
+        self.statistics_ = np.ma.array(X, mask=mask).mean(axis=0).data
+
+        return self
+
+    def transform(self, X):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : array-like, shape = [n_samples, n_features]
+            The input data to complete.
+
+        Returns
+        -------
+        X : array-like, shape = [n_samples, n_features]
+            The imputed dataset.
+        """
+
+        check_is_fitted(self, ["fit_X_", "statistics_"])
+        if not is_scalar_nan(self.missing_values):
+            force_all_finite = True
+        else:
+            force_all_finite = "allow-nan"
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+
+        if X.shape[1] != self.fit_X_.shape[1]:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed")
+
+        mask = _get_mask(X, self.missing_values)
+        orig_X_shape = X.shape
+        if not np.any(mask):
+            return X
+
+        # Check for excessive missingness in rows
+        row_total_missing = mask.sum(axis=1)
+        bad_rows = row_total_missing > (mask.shape[1] *
+                                        self.feature_max_missing)
+        if np.any(bad_rows):
```

Comment:
Is it faster to do this than to just compute them? Or is the point that they might all be NaN and then we get an error below?